source/slang/slang-ir-inline.cpp

```diff
@@ -355,6 +355,52 @@ struct InliningPassBase
         return true;
     }
 
+        // When instructions are cloned, with cloneInst no sourceLoc information is copied over by default.
+        // Here we attempt some policy about copying sourceLocs when inlining.
+        //
+        // An assumption here is that [__unsafeForceInlineEarly] will not be in user code (when we have more
+        // general inlining this will not follow).
+        //
+        // Therefore we probably *don't* want to copy sourceLoc from the original definition in the stdlib because
+        //
+        // * That won't be much use to the user (they can't easily see stdlib code currently for example)
+        // * That the definitions in stdlib are currently 'mundane' and largely exist to flesh out language features - such that
+        //   their being in the stdlib would likely be surprising to users
+        //
+        // That being the case, we actually copy the call sites sourceLoc if it's defined, and only fall back
+        // onto the originating loc, if that's not defined.
+        //
+        // We *could* vary behavior if we knew if the function was defined in the stdlib. There doesn't appear 
+        // to be a decoration for this.
+        // We could find out by looking at the source loc and checking if it's in the range of stdlib - this would actually be
+        // a fast and easy but to do properly this way you'd want a way to mark that source range that would also work across
+        // serialization.
+        // 
+        // For now this punts on this, and just assumes [__unsafeForceInlineEarly] is not in user code.
+    static IRInst* _cloneInstWithSourceLoc(CallSiteInfo const& callSite,
+        IRCloneEnv*     env,
+        IRBuilder*      builder,
+        IRInst*         inst)
```

Comment:
This function seems like a good solution to the problem, and my only question is whether we can make it something more integrated/systematic.

One obvious tweak would be to make the `cloneInst` operation default to copying the `sourceLoc` of the instruction being cloned (which makes sense for a "clone" operation anyway). That would mean that the `else if` bit below is not needed.

Another potentially interesting bit would be to have the `IRBuilder` type include a field called something like `sourceLocToUseForOverride` that can optionally be set and, when it is set, becomes the source location of all instructions emitted for that `IRBuilder`. That would mean that the `inlineTrivialFuncBody()` below could just set up the "override location" based on the `callSite` and allow other machinery to do the rest of the work.